posts/2024-02-15-waku-network-tech-overview.mdx

```diff
@@ -0,0 +1,178 @@
+
+### Message Routing - Autosharding
+
+One issue with sharding is that users and applications need to know what shard to use.
+
+One application could arbitrarily decide what shard they want to use but that may not scale easily, especially when adding new shards. It is also one more decision for the developer to make; we prefer to make the dev ex as easy as possible. It also gives the opportunity for developers to build applications that spread on several shards without delegating shard choices to the user.
+
+Autosharding is a simple protocol that dispatches messages on a shard based on the application.
+
+### Message Routing - RLN Relay
+
+Waku is generalized, meaning any kind of payload can be transported. Hence, there is no strict definition of “spam”. Whether the message contains a [meme](https://github.com/b4s36t4/waku-meme) or a [zk note for private DeFi](https://railgun.ch/#/), Waku shouldn't even know the content; metadata and data should remain private.
+
+Hence, there is a risk of someone pushing gigabits of data to the network. This can be a problem at different levels:
+
+- Bandwidth usage: this may lead to a user’s bandwidth being hogged and affecting other services (streaming, staking), or having a surprise bill.
+- Connectivity: if a node does not have enough bandwidth to send/receive all messages on the gossipsub layer, then their behavior may be seen as incorrect from other nodes that may disconnect from it.
+- Reliability: If the traffic is higher than the available bandwidth, then a node may not be able to reliably send and receive messages.
+- Other resources: memory usage is correlated with the traffic, so is disk space for a store service.
+
+So rather than defining what a "spam" message looks like, we introduced rate limit on the network to enable a fair use of Waku, with a capping of the bandwidth usage per shard.
+
+This is done with RLN, or Rate Limiting Nullifier, which enables to limit the rate of messages sent by a given publisher.
+It is currently set to 1 msg/s. Coupled with a maximum message size (150kB) and a maximum number of publishers (80k, TBD), we are able to assume a maximum bandwidth usage per shard (around 10Mbps).
+
+Rate-limiting publishers in a censorship-resistant and private manner is difficult; this is why we are using zero-knowledge technology:
+
+1. Users push their RLN credentials to a smart contract (currently on Ethereum Sepolia testnet).
+2. Nodes watch the contract and include the new membership.
+3. When sending a message, the user attaches an RLN proof to the message with the current epoch (timestamp in seconds).
+4. Nodes can verify the proof, without being able to correlate the user’s Ethereum address (used on the smart contract) with the message being sent, preserving anonymity.
+5. If a user attempts to send more than 1 msg/s, nodes can detect this and drop the message surplus, or spam, protecting the network.
```

Comment:
Importantly: and disconnect from peers forwarding spam, thereby isolating the spam origin from the network.